### PR TITLE
`eqx.filter_shard`; test + update `examples/parallelism.ipynb`

### PR DESCRIPTION
**Pull request by homerjed**
Hi Patrick,

Please pardon the terrible `git` practice if I've made some mistakes. I tend to work alone in my daily projects :) 

I've

- added `equinox/_sharding.py` with `eqx.filter_shard` as per the issue thread from the other day,
- added a test `tests/test_sharding.py` that tests the function on sharding an `eqx.nn.MLP` in and outside of `eqx.filter_jit`,
- changed the `examples/parallelism.ipynb` example to have the same content as the same file in your PR you weren't sure about, with the corresponding changes made.

Let me know what you think, again, sorry if the `git` stuff is wrong!

Cheers,
Jed

### CODE REVIEW

File: equinox/_sharding.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    single-device sharding (`jaxlib.xla_extension.Device`).
+    **Arguments:**
+    - `x`: A PyTree, with potentially a mix of arrays and non-arrays on the leaves.
+        They will have their shardings constrained.
+    - `device_or_shardings`: Either a singular device (e.g. CPU or GPU) or PyTree of
+        sharding specifications. The structure should be a prefix of `x`.
+    **Returns:**
+    A copy of `x` with the specified sharding constraints.
+    !!! Example
+        See also the [autoparallelism example](../../../examples/parallelism).
```
[the remark this answers is not in the document]
Comment (homerjed) [reply equinox/_sharding.py]: I missed this, I've shortened the docstring a bit and formatted it correctly, looks less horrific on `mkdocs serve`.

File: tests/test_sharding.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+
+def test_sharding():
+    mlp = eqx.nn.MLP(2, 2, 2, 2, key=jr.PRNGKey(0))
+
+    eqx.filter_shard(mlp, cpu)
+
+    @eqx.filter_jit
+    def f(x):
+        return eqx.filter_shard(x, sharding)
```
[the remark this answers is not in the document]
Comment (homerjed) [reply tests/test_sharding.py:20]: Fixed the test here to actually do something, basically just filters the params out and adds one to them, then `filter_shard`s it.

**Comment by patrick-kidger** [comment]
Alright, this LGTM! That is, modulo David's concern back in #688. Let's double check with things over there before we merge this.

**Comment by patrick-kidger** [comment]
Okay, looks like the discussion has settled, so merging! Thank you for putting this together, and I'm excited to have this in the next release of Equinox! :)

**Comment by homerjed** [comment]
Cheers Patrick! Hope to contribute again :)

**Pull request merged**